### PR TITLE
[Tables] use etag from entity if match condition is given

### PR DESCRIPTION
**Pull request by seankane-msft**

### CODE REVIEW

File: sdk/tables/azure-data-tables/azure/data/tables/_table_batch.py [hunk 1]
```diff
@@ -238,11 +238,19 @@ def update(
         self._verify_partition_key(entity)
         temp = entity.copy()
 
+        match_condition = kwargs.pop("match_condition", None)
+        etag = kwargs.pop("etag", None)
+        if match_condition and not etag:
+            try:
+                etag = entity.metadata.get("etag", None)
+            except AttributeError:
```
Comment (annatisch) [inline sdk/tables/azure-data-tables/azure/data/tables/_table_batch.py:246]: I think you might also need to catch `TypeError` here - if a customer passes in a `TableEntity` instance they created, then `entity.metadata` will be `None` (and we should add a test for this).

File: sdk/tables/azure-data-tables/azure/data/tables/_table_client.py [hunk 2]
```diff
@@ -414,12 +421,16 @@ def update_entity(
                 :dedent: 8
                 :caption: Updating an already exiting entity in a Table
         """
+        match_condition = kwargs.pop("match_condition", None)
+        etag = kwargs.pop("etag", None)
+        if match_condition and not etag:
+            etag = entity.metadata.get("etag")
```
Comment (annatisch) [inline sdk/tables/azure-data-tables/azure/data/tables/_table_client.py:427]: This one's missing the try/except?

File: sdk/tables/azure-data-tables/azure/data/tables/aio/_table_client_async.py [hunk 3]
```diff
@@ -402,11 +411,26 @@ async def update_entity(
                 :dedent: 8
                 :caption: Querying entities from a TableClient
         """
+        match_condition = kwargs.pop("match_condition", None)
+        etag = kwargs.pop("etag", None)
+        if match_condition and not etag:
+            etag = entity.metadata.get("etag")
```
Comment (annatisch) [inline sdk/tables/azure-data-tables/azure/data/tables/aio/_table_client_async.py:417]: Also missing try/except here

File: sdk/tables/azure-data-tables/azure/data/tables/aio/_table_client_async.py [hunk 4]
```diff
@@ -402,11 +411,26 @@ async def update_entity(
...
+
+        if_match, _ = _get_match_headers(
```
Comment (annatisch) [inline sdk/tables/azure-data-tables/azure/data/tables/aio/_table_client_async.py:419]: Seems like we duplicated this section, as it's also on line 429?

File: sdk/tables/azure-data-tables/tests/test_table_entity.py [hunk 5]
```diff
@@ -898,23 +898,46 @@ def test_get_entity_if_match(self, tables_storage_account_name, tables_primary_s
...
+
+            entity["value"] = 10
+            self.table.update_entity(entity)
+
+            # Get Entity and set old etag
+            e = self.table.get_entity(entity["PartitionKey"], entity["RowKey"])
+            new_etag = e.metadata["etag"]
+            e.metadata["etag"] = old_etag
+
+            with pytest.raises(HttpResponseError):
```
Comment (annatisch) [inline sdk/tables/azure-data-tables/tests/test_table_entity.py:935]: Is this actually just a raw `HttpResponseError`? Or are we raising `azure.core.ResourceModifiedError`?

Comment (seankane-msft) [reply sdk/tables/azure-data-tables/tests/test_table_entity.py:935]: This is a raw `HttpResponseError`, here's what it looks like.

```
E           azure.core.exceptions.HttpResponseError: The update condition specified in the request was not satisfied.
E           RequestId:0c9fe685-8002-003d-4cd4-41ac3f000000
E           Time:2021-05-05T17:33:24.9865841Z
E           ErrorCode:UpdateConditionNotSatisfied
```

File: sdk/tables/azure-data-tables/azure/data/tables/_table_batch.py (hunk 1 above)
Comment (seankane-msft) [reply sdk/tables/azure-data-tables/azure/data/tables/_table_batch.py]: Updated and added a test!

File: sdk/tables/azure-data-tables/tests/test_table_entity.py (hunk 5 above)
Comment (annatisch) [reply sdk/tables/azure-data-tables/tests/test_table_entity.py:935]: Interesting! I've made a note to fix that - but could come post-GA if needed.

**Review by annatisch — APPROVED** [review]

**Pull request merged**